Titles : Migrate Add & Edit Forms

Progi1984 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/8/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Titles : Migrate Add & Edit Forms
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Relative to #10629
| How to test?      | Enable the Feature Flag "Titles" : The page Titles is now migrated. Check behavior for Listing / Add / Edit / Delete / Bulk.


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

Progi1984 committed:
Titles : Migrate Add & Edit Forms

yanmakouf reviewed (dismissed):
Everything looks great to me :)

Progi1984 commented:
Ping @PrestaShop/prestashop-maintainers & @PrestaShop/committers for a review

zuk3975 commented on src/Adapter/Title/CommandHandler/AddTitleHandler.php:
```diff
@@ -0,0 +1,99 @@
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Title\CommandHandler;
```
ping @jolelievre should we stop using the Adapter ns already? (since the ADR https://github.com/PrestaShop/ADR/blob/master/0014-drop-the-no-use-of-legacy-in-new-code-restriction.md)

zuk3975 reviewed (commented):
The whole PR seems to follow previous migration examples with abstract handlers etc.. and maybe thats ok?
I don't know if we should request to implement repository and validator services as I asked here https://github.com/PrestaShop/PrestaShop/pull/28812#discussion_r1020008018 @jolelievre ?
Repository and validators have its own benefits and are quite easy to implement, plus they are mentioned in the devdocs, so I would want to see them in action, but I don't want to be the bad guy forcing them in a PR that is already done and functioning well (even though its based on abit older examples) :man_shrugging:

Progi1984 commented:
@PrestaShop/committers Hi team, when you have some available time, could you review this PR, plz ?

FabienPapet commented on src/Adapter/Title/AbstractTitleHandler.php:
```diff
@@ -0,0 +1,107 @@
+        int $imageWidth,
+        int $imageHeight
+    ): void {
+        $temporaryImage = tempnam(_PS_TMP_IMG_DIR_, 'PS');
```
use DI to inejct constant

Progi1984 replied:
:heavy_check_mark:

FabienPapet commented on src/Adapter/Title/AbstractTitleHandler.php:
```diff
@@ -0,0 +1,107 @@
+        // Copy new image
+        if (!ImageManager::resize(
+            $temporaryImage,
+            _PS_IMG_DIR_ . $imageDir . $titleId . '.jpg',
```
same usage of constant here

FabienPapet replied:
Fixed while submitting the review.

Progi1984 replied:
:heavy_check_mark:

FabienPapet commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/CustomerPreferences/TitleType.php:
```diff
@@ -0,0 +1,99 @@
+
+namespace PrestaShopBundle\Form\Admin\Configure\ShopParameters\CustomerPreferences;
+
+use Gender;
```
Usage of legacy here, please use a constant inside the `Core` namespace.

FabienPapet replied:
Fixed while submitting the review.

Progi1984 replied:
Weird... it's a new architecture rule... I didn't see it.

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/form/form_data_handler.yml:
```diff
@@ -263,3 +263,7 @@ services:
       - '@prestashop.core.command_bus'
+
+  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataHandler\TitleFormDataHandler:
+    arguments:
+      - '@prestashop.core.command_bus'
```
```suggestion
    autowire: true
```

Progi1984 replied:
:heavy_check_mark:

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/form/form_data_provider.yml:
```diff
@@ -228,3 +228,7 @@ services:
       - '@prestashop.core.query_bus'
+
+  PrestaShop\PrestaShop\Core\Form\IdentifiableObject\DataProvider\TitleFormDataProvider:
+    arguments:
+      - '@prestashop.core.query_bus'
```
```suggestion
    autowire: true
```

Progi1984 replied:
:heavy_check_mark:

Progi1984 committed:
Surprise : it's not anymore allowed to have legacy constants classes in Core namespace

FabienPapet commented on src/Core/Domain/Title/Exception/TitleConstraintException.php:
```diff
@@ -32,4 +32,7 @@
 class TitleConstraintException extends TitleException
 {
     public const INVALID_ID = 1;
+
```
This is wierd that phpcsfixer does not complain about the extra line here, or the missing line l37

Progi1984 replied:
:heavy_check_mark:

Progi1984 committed:
Use DI for constants, and autowire for services

Hlavtox commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/CustomerSettings/Title/Blocks/form.html.twig:
```diff
@@ -0,0 +1,50 @@
+        </div>
+      </div>
+      <div class="card-footer">
+        <a href="{{ path('admin_contacts_index') }}" class="btn btn-outline-secondary">
```
Bad link, should be `admin_title_index`

Progi1984 replied:
Fixed

Hlavtox commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/CustomerSettings/Title/Blocks/form.html.twig:
```diff
@@ -0,0 +1,50 @@
+    <div class="card">
+      <h3 class="card-header">
+        <i class="material-icons">man</i>
+        {{ 'Social titles'|trans({}, 'Admin.Shopparameters.Feature') }}
```
`Social title` or `Title`, we are editing a single entity.

Progi1984 replied:
Ping @l-delin for confirmation

Progi1984 replied:
Context : It's the page "Add a new title"

l-delin replied:
Hello! 
We should use "Title" as we only refer to using Mr. Mrs. (see screenshot) and not to other titles [(Lord, Count, Duke...)](https://en.wikipedia.org/wiki/Category:Men%27s_social_titles):
<img width="724" alt="image" src="https://user-images.githubusercontent.com/113511191/221541128-331424f5-580a-4294-b083-09addfb26dba.png">

Following those examples, `Social title` doesn't make a lot of sense here, so we should use `Title`. What do you think, @Hlavtox ?

Hlavtox replied:
@l-delin I think we should use `Title`.

From a quick google I think that:
- Gender: Male, Female
- Title: Mr., Mrs., Professor, Sir

l-delin replied:
Let's change it to `Title`, then 🎉

Hlavtox commented:
Can we add user-friendly translatable versions to exceptions that can occur in forms?

Progi1984 committed:
Titles : Fixed path

Progi1984 commented:
@Hlavtox & @FabienPapet : Fixed the error in path in ae17108. Please review :D

zuk3975 commented on src/Adapter/Title/CommandHandler/AddTitleHandler.php:
```diff
@@ -0,0 +1,83 @@
+     */
+    public function handle(AddTitleCommand $command): TitleId
+    {
+        $title = new Gender();
```
why Gender entity is referred as title everywhere? I find it so confusing :cold_sweat:

Progi1984 replied:
It's the new term for replacing Gender, so I apply it before deprecations.

zuk3975 commented:
I noticed routing is not following our conventions.
it should be /titles (plural), I see it was probably related to your previous PR, maybe you could fix that in current (or dedicated) PR?

Progi1984 committed:
Titles : Rename the URL to the plural

Progi1984 commented:
> I noticed routing is not following our conventions. https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Resources/config/routing/admin/configure/shop_parameters/_shop_parameters.yml#L47 it should be /titles (plural), I see it was probably related to your previous PR, maybe you could fix that in current (or dedicated) PR?

:heavy_check_mark: in 68907e9

l-delin commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/CustomerPreferences/TitleType.php:
```diff
@@ -0,0 +1,99 @@
+        $builder
+            ->add('name', TranslatableType::class, [
+                'label' => $this->trans(
+                    'Social title',
```
```suggestion
                    'Title',
```

l-delin commented on src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/CustomerSettings/Title/Blocks/form.html.twig:
```diff
@@ -0,0 +1,50 @@
+    <div class="card">
+      <h3 class="card-header">
+        <i class="material-icons">man</i>
+        {{ 'Social titles'|trans({}, 'Admin.Shopparameters.Feature') }}
```
```suggestion
        {{ 'Titles'|trans({}, 'Admin.Shopparameters.Feature') }}
```

l-delin reviewed (commented):
![image](https://user-images.githubusercontent.com/113511191/221562091-30fe76e0-c832-490a-8db3-8e2bac2a4b01.png)
![image](https://user-images.githubusercontent.com/113511191/221562167-f7931a4c-7d38-493c-a3d6-c4e790406410.png)

Progi1984 committed:
Titles : Fixed wording

Progi1984 commented:
@Hlavtox & @zuk3975 I fixed the wording for @l-delin in the last commit 324b8b26340d1ebc9146ad5cb684a96f410bbff6.

djoelleuch reviewed (commented):
Hello @Progi1984  , 

i found some issues : 

**Issue 1 :**  

1. Add new title > fill all the blanks >click on save .
2. edit the title created > see the image ,image width ,image height are not saved !

[Untitled_ Feb 27, 2023 11_40 AM.webm](https://user-images.githubusercontent.com/115152460/221578249-7ae2d931-2969-4a68-8bf8-750e1f06703d.webm)

**Issue 2** : 

1. Add new title > fill all the blanks >change the language to french >see nothing is changing !

[Untitled_ Feb 27, 2023 11_43 AM.webm](https://user-images.githubusercontent.com/115152460/221579088-459f612a-f94c-4fbd-88db-e2c4a62aa8d6.webm)

**issue 3 :**  :

1.  click on add new title > the new title page is well display but the name of the page is NOK
_the migrate page_ 
![image](https://user-images.githubusercontent.com/115152460/221587578-fdb49f06-1a07-49cb-b459-1515d6c64c99.png)
_the old page_ 
![image](https://user-images.githubusercontent.com/115152460/221587948-aa7f7fc0-e56d-4078-bfb6-e03300efe4c8.png)
Could you check please ?

Hlavtox commented:
@djoelleuch 
1. The image sizes are used only for uploading the image, there is no such field in the database.
2. Lang switcher not initialized or assets not built - @Progi1984?
3. The titles are a mess, but I think it could be handled in a separate PR we are working on - https://github.com/PrestaShop/PrestaShop/pull/31322

Progi1984 committed:
Titles : Fixed language selector

Progi1984 committed:
Titles : Set default image width/height

Progi1984 commented:
@djoelleuch / @Hlavtox 
1. I fixed and kept the same behavior than in legacy pages.
2. Fixed
3. Fixed after the feedback from @l-delin 

Need two reviews from @Hlavtox / @zuk3975 / @PrestaShop/committers / @PrestaShop/prestashop-maintainers

Progi1984 commented:
Ping @committers / @prestashop-maintainers for review

zuk3975 commented on src/Adapter/Title/AbstractTitleHandler.php:
```diff
@@ -0,0 +1,115 @@
+     * @var string
+     */
+    protected $tmpImageDir;
+    /**
```
missing extra lines between properties. But should be spotted with cs fixer ideally. Not blocking
```suggestion

    /**
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Adapter/Title/CommandHandler/DeleteTitleHandler.php:
```diff
@@ -28,27 +28,22 @@
+        $title = $this->titleRepository->get($command->getTitleId());
 
         if (!$title->delete()) {
             throw DeleteTitleException::createDeleteFailure($command->getTitleId());
```
handle method should be just one line
```suggestion
            $this->titleRepository->delete($command->getTitleId());
```
in repository you can reuse get(TitleId $titleId) and implement new delete(TitleId $titleId) method reusing AbstractObjectModelRepository::deleteObjectModel

Progi1984 replied:
Fixed

zuk3975 commented on src/Adapter/Title/Repository/TitleRepository.php:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * @param Gender $title
+     */
+    public function update(Gender $title): void
```
is this method used? 
In update handler I see you use partialUpdate, so unless it is used somewhere, you can get rid of it

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/ValueObject/Gender.php:
```diff
@@ -0,0 +1,36 @@
+{
+    public const TYPE_MALE = 0;
+    public const TYPE_FEMALE = 1;
+    public const TYPE_OTHER = 2;
```
missing constructor which validates the type. Then you can use this class as input for commands

zuk3975 replied:
and a gettter `getValue(): int`

zuk3975 replied:
and then don't forget to adjust the handlers (`$command->getGender()->getValue()`)

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/AddTitleCommand.php:
```diff
@@ -0,0 +1,121 @@
+    protected $localizedNames;
+
+    /**
+     * @var int
```
use your new VO
```suggestion
     * @var Gender
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/AddTitleCommand.php:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * @var int
+     */
+    protected $genderType;
```
might want to rename to $gender

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/AddTitleCommand.php:
```diff
@@ -0,0 +1,121 @@
+        ?int $imgHeight
+    ) {
+        $this->localizedNames = $localizedNames;
+        $this->genderType = $genderType;
```
```suggestion
        $this->genderType = new Gender($gender);
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/AddTitleCommand.php:
```diff
@@ -0,0 +1,121 @@
+     */
+    public function __construct(
+        array $localizedNames,
+        int $genderType,
```
```suggestion
        int $gender,
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/EditTitleCommand.php:
```diff
@@ -0,0 +1,186 @@
+    /**
+     * @return int|null
+     */
+    public function getGenderType(): ?int
```
```suggestion
    public function getGenderType(): ?Gender
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/EditTitleCommand.php:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    /**
+     * @param int $genderType
```
```suggestion
     * @param int $gender
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/EditTitleCommand.php:
```diff
@@ -0,0 +1,186 @@
+     *
+     * @return self
+     */
+    public function setGenderType(int $genderType): self
```
```suggestion
    public function setGender(int $gender): self
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/EditTitleCommand.php:
```diff
@@ -0,0 +1,186 @@
+     */
+    public function setGenderType(int $genderType): self
+    {
+        $this->genderType = $genderType;
```
```suggestion
        $this->genderType = new Gender($gender);
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/QueryResult/EditableTitle.php:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * @var int
+     */
+    protected $genderType;
```
can also be renamed to gender (just to stay consistent), but value should stay `int`, because we don't use value objects in query results, so its ok :+1: 
```suggestion
    protected $gender;
```

Progi1984 replied:
Fixed

zuk3975 commented on tests/Integration/Behaviour/Features/Scenario/Title/title_management.feature:
```diff
@@ -11,21 +11,15 @@ Feature: Title management
+
+  Scenario: Editing title
+    When I edit title "3rdTitle" with following properties:
+      | name    | Fourth Title   |
```
could I ask to test multilang?
Plenty of examples under product domain, but you would basically need to add couple language assertions steps in background:
```
    And language "en" with locale "en-US" exists
    And language "fr" with locale "fr-FR" exists
    And language with iso code "en" is the default one
```

and then when editing/creating use it like this to provide localized langs
```suggestion
      | name[en-US]  | Fourth Title             |
      | name[fr-FR]    | Fourth Title  french  |
```

and in the context file you can use $this->localizeByRows($tableNode) and you will get prepared names indexed by id

Progi1984 replied:
Fixed

zuk3975 commented on src/Adapter/Title/AbstractTitleHandler.php:
```diff
@@ -0,0 +1,115 @@
+     * @param int $imageWidth
+     * @param int $imageHeight
+     */
+    protected function uploadImage(
```
I think you could use PrestaShop\PrestaShop\Core\Image\Uploader\ImageUploaderInterface instead of this method.
Check examples of that interface implementations

Progi1984 replied:
Fixed

zuk3975 commented on src/Adapter/Title/CommandHandler/EditTitleHandler.php:
```diff
@@ -0,0 +1,91 @@
+            return;
+        }
+
+        $title->deleteImage();
```
this method should also be moved to repository and wrapped into try-catch block throwing CoreException only (so that legacy Prestashop exceptions doesn't bubble up), smth like
```
try {
    $title->deleteImage();
catch (PrestashopException $e) {
    throw new CoreException(sprintf('Error occurred when trying to delete title %d image', $title->id));
}
```

Progi1984 replied:
Fixed

zuk3975 commented on src/Core/Domain/Title/Command/AddTitleCommand.php:
```diff
@@ -0,0 +1,121 @@
+class AddTitleCommand
+{
+    /**
+     * @var array<string>
```
```suggestion
     * @var array<int, string>
```

Progi1984 replied:
Fixed

Progi1984 committed:
Titles : Fixed feedbacks : behat tests (multilang), exit the uploader, value object the gender

Progi1984 commented:
@Hlavtox & @zuk3975 : Feedbacks adressed :muscle:

jolelievre commented on src/Adapter/Image/Uploader/TitleImageUploader.php:
```diff
@@ -0,0 +1,115 @@
+     * @throws UploadedImageConstraintException
+     * @throws ImageUploadException
+     */
+    public function upload($titleId, UploadedFile $image)
```
```suggestion
    public function upload($titleId, UploadedFile $image, int $imageWidth, int $imageHeight)
```
Why not specify the sizes as parameters, this would prevent making this service stateful, especially since all the usages so far always call the two setters, then the upload method. So maybe only `upload` could be enough as a public method.

jolelievre replied:
Oh I think I get it! That's because of the `ImageUploaderInterface` right?
I believe that you could add two optional parameters without breaking the interface, that's actually been done a few times with implementations of `UrlProviderInterface` that needed some parameters.

That said it becomes not so much interesting to use an interface in this case if it doesn't meet the need, which kind of make me wonder if you should implement this interface here, even `AbstractTitleHandler` doesn't rely on the interface but the implementation itself. Maybe extending `AbstractImageUploader` is enough?

jolelievre commented on src/Adapter/Title/CommandHandler/AddTitleHandler.php:
```diff
@@ -0,0 +1,83 @@
+            return;
+        }
+
+        $this->titleImageUploader->setImageHeight($command->getImageHeight());
```
I'm wondering what happens if `getImageHeight` and/or `getImageWidth` are null (as they are optional), should the three parameters be assigned altogether for this upload to work? If so it's fine, then maybe the command should check that the three parameters are either all set or none at all?

Or you can rely on a setter based on three parameters, this way it always ensures that the three parameters are set at the same time.

jolelievre replied:
I actually checked the legacy page and those two size fields are kind of weird, since they only make sense at the moment you upload the image.. Not sure if the UX is ideal here 😅 But that can be discussed with @MatShir afterwhile let's stick with the iso functional for now.

So I was just thinking maybe as a fallback you should use the `TitleSettings` const here when `getImageHeight` or `getImageWidth` is null

jolelievre commented on src/Adapter/Title/CommandHandler/BulkDeleteTitleHandler.php:
```diff
@@ -28,28 +28,23 @@
-            if (0 >= $title->id) {
-                throw new TitleNotFoundException(sprintf('Unable to find title with id "%d" for deletion', $titleId->getValue()));
-            }
+            $title = $this->titleRepository->get($titleId);
```
```suggestion
            $this->titleRepository->delete($this->titleRepository->get($titleId));
```

jolelievre replied:
Only this one line should cover it all even the exception throwing

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Domain/Title/QueryResult/EditableTitle.php:
```diff
@@ -0,0 +1,91 @@
+     * @param Gender $gender
+     */
+    public function __construct(
+        TitleId $titleId,
+        array $localizedNames,
+        Gender $gender
```
```suggestion
        int $titleId,
        array $localizedNames,
        int $gender
```

jolelievre replied:
Sorry, but in our convention, we don't use VOs in query results but only scalar values (except for DecimalNumber). I don't remember exactly all the reasons but amongst them the idea is that in commands VOs are used partially to prevent invalid inputs (the input in command should be scalar though), but in query result though since the model is supposed to be consistent there can be no invalid values.

And also having mostly scalar values should ease the serialization.

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Form/IdentifiableObject/DataHandler/TitleFormDataHandler.php:
```diff
@@ -0,0 +1,97 @@
+        /** @var TitleId $titleId */
+        $titleId = $this->commandBus->handle(new AddTitleCommand(
+            $data['name'],
+            new Gender($data['genderType']),
```
```suggestion
            $data['genderType'],
```
The input parameters of command should be scalar inputs, but they are then converted in the constructor as VOs

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Domain/Title/Command/AddTitleCommand.php:
```diff
@@ -0,0 +1,124 @@
+     */
+    public function __construct(
+        array $localizedNames,
+        Gender $gender,
```
We try to use scalar inputs as much as possible in input parameters for commands, but they are instantly converted into VOs, there are a few exceptions but only when a simple scalar is not enough like:
- ShopConstraint
- UploadedFile
- DecimalNumber

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Domain/Title/Command/EditTitleCommand.php:
```diff
@@ -0,0 +1,188 @@
+     *
+     * @return self
+     */
+    public function setGender(Gender $gender): self
```
```suggestion
    public function setGender(int $gender): self
```
Same suggestion here

Progi1984 replied:
Fixed

jolelievre commented on src/Core/Domain/Title/Command/EditTitleCommand.php:
```diff
@@ -0,0 +1,188 @@
+     *
+     * @return self
+     */
+    public function setGender(Gender $gender): self
```
```suggestion
    public function setGender(int $gender): self
```
Same suggestion here

Progi1984 replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/CustomerPreferences/TitleType.php:
```diff
@@ -0,0 +1,99 @@
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Image width in pixels. Enter "0" to use the original size.',
```
Hmm, so the fallback when no size is specified is to keep the original size? That didn't seem to be the case in legacy page but why not

Progi1984 replied:
It was the case in legacy page : https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminGendersController.php#L172-L185

jolelievre commented on src/PrestaShopBundle/Resources/config/services/adapter/title.yml:
```diff
@@ -11,12 +29,34 @@ services:
     class: 'PrestaShop\PrestaShop\Adapter\Title\CommandHandler\DeleteTitleHandler'
+    autowire: true
+    arguments:
+      $tmpImageDir: !php/const _PS_TMP_IMG_DIR_
+      $genderDir: !php/const _PS_GENDERS_DIR_
```
Isn't it weird to inject these values even in handlers that don't need them just because they all extend the same abstract handler?

Progi1984 replied:
Removed. Thanks ;)

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/TitleFeatureContext.php:
```diff
@@ -28,14 +28,109 @@
+            $titleId = $this->getCommandBus()->handle(new AddTitleCommand(
+                $data['name'],
+                new Gender((int) $availableGendersTypes[$data['type']]),
+                null,
+                null,
+                null
```
If these three parameters are not mandatory maybe they should be handled by a dedicated setter (for the three parameters), or at least the should be nullable in the constructor so we don't need to always set null explicitely

Progi1984 replied:
Fixed

jolelievre reviewed (commented):
Thanks @Progi1984 

I have a few comments and questions on a few files, but all the other are good!
Also one question do you think it's worth testing the image upload in the behat tests?

Progi1984 committed:
Titles : Delete improvements

Progi1984 committed:
Titles : In Query, use only scalar parameters

Progi1984 committed:
Titles : In Command, use only scalar parameters

Progi1984 committed:
Titles : Remove unuseful parameters

Progi1984 committed:
Titles : Set default values for AddTitleCommand

Progi1984 committed:
Titles : Improved TitleImageUploader method

Progi1984 committed:
Titles : Fixed PHPCSFixer

zuk3975 commented on src/Adapter/Title/CommandHandler/BulkDeleteTitleHandler.php:
```diff
@@ -28,32 +28,24 @@
-            if (!$title->delete()) {
-                throw DeleteTitleException::createBulkDeleteFailure($titleId);
-            }
+            $this->titleRepository->delete(
+                $this->titleRepository->get($titleId)
+            );
```
could have been 
`$this->titleRepository->delete($titleId)` 
and the repository would get the obj model inside.
But its not blocking for me, because recently in some other PR I noticed if the deletion is more complicated (e.g if you would need to check some Title fields before deletion) you would still need to fetch the obj model before so this way would make more sense :man_shrugging:. So idk, both ways are good for me, not sure which is better globally :smile:

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/CustomerPreferences/TitleType.php:
```diff
@@ -0,0 +1,99 @@
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('name', TranslatableType::class, [
```
default language constraint missing I think
```suggestion
            ->add('name', TranslatableType::class, [
                'constraints' => [new DefaultLanguage()]
```

Progi1984 replied:
Fixed

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/CustomerPreferences/TitleType.php:
```diff
@@ -0,0 +1,99 @@
+                'expanded' => true,
+                'placeholder' => false,
+                'choices' => [
+                    $this->trans('Male', 'Admin.Shopparameters.Feature') => Gender::TYPE_MALE,
+                    $this->trans('Female', 'Admin.Shopparameters.Feature') => Gender::TYPE_FEMALE,
+                    $this->trans('Neutral', 'Admin.Shopparameters.Feature') => Gender::TYPE_OTHER,
```
would say this could go to choice provider (at least based on how we have done it before), but recently @jolelievre and @FabienPapet had some discussions about maybe leaving choices in related forms, so Im not requesting that change :smile:

zuk3975 reviewed (approved):
just couple side notes, but I approve

zuk3975 reviewed (changes requested):
Oh, the DefaultLanguage constraint is missing, thats a must

Progi1984 committed:
Titles : Added DefaultLanguage Constraint for the form

jolelievre commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/CustomerPreferences/TitleType.php:
```diff
@@ -0,0 +1,101 @@
+                    'Admin.Global'
+                ),
+            ])
+            ->add('genderType', ChoiceType::class, [
```
```suggestion
            ->add('gender_type', ChoiceType::class, [
```
We don't use camel case usually for form fields

Progi1984 replied:
Fixed

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/TitleFeatureContext.php:
```diff
@@ -28,14 +28,106 @@
 
 class TitleFeatureContext extends AbstractDomainFeatureContext
 {
+    /**
+     * @var int default lang id from configs
+     */
+    protected $defaultLangId;
+
+    public function __construct()
+    {
+        $configuration = CommonFeatureContext::getContainer()->get('prestashop.adapter.legacy.configuration');
+        $this->defaultLangId = (int) $configuration->get('PS_LANG_DEFAULT');
+    }
```
```suggestion
```
You don't need all this as you can use the `AbstractDomainFeatureContext::getDefaultLangId` protected method already

Progi1984 replied:
Better. The property is not used. So I remove it.

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/TitleFeatureContext.php:
```diff
@@ -79,4 +171,107 @@ public function assertLastErrorTitleNotFound(): void
+    /**
+     * @Then the title :titleReference name should be :name in locale :locale
+     */
+    public function assertTitleName(string $titleReference, string $name, string $locale): void
```
Asserting values sentence by sentence is not convenient at all. Usually we use a table with some associative columns, based on `TableNode` It should have the same format as the add/edit commands input format so it can be copied easily into feature files

jolelievre replied:
For localized values there is even a `localizeByRows` (that you already used for the commands) that takes care of converting `name[fr-FR]` into localized data with the language IDs already adapted

Progi1984 replied:
Fixed

jolelievre commented on tests/Integration/Behaviour/Features/Scenario/Title/title_management.feature:
```diff
@@ -6,26 +6,26 @@ Feature: Title management
+      | name[en-US] | Fourth Title EN |
+      | name[fr-FR] | Fourth Title FR |
+      | type        | Female          |
+    Then the title "3rdTitle" name should be "Fourth Title EN" in locale "en-US"
+    Then the title "3rdTitle" name should be "Fourth Title FR" in locale "fr-FR"
+    And the title "3rdTitle" gender type should be "Female"
```
As I suggested in the context class this should all be tested in a single step with data in table like this:

```suggestion
    Then the title "3rdTitle" should have the following properties:
      | name[en-US] | Fourth Title EN |
      | name[fr-FR] | Fourth Title FR |
      | type        | Female          |
```

Progi1984 replied:
Fixed

jolelievre reviewed (commented):
Thanks @Progi1984 

Last few details and improvement in behat tests and it's all good!

Progi1984 committed:
Titles : Changed key (not in camelCase) in form

Progi1984 committed:
Titles : Behat changes for checking all properties in one shot

djoelleuch reviewed (approved):
Hello @Progi1984 
LGTM , the issues are fixed :heavy_check_mark: 

![image](https://user-images.githubusercontent.com/115152460/232530802-5e55edd5-a1af-4aaf-98c1-440160befb4a.png)

[Untitled_ Apr 17, 2023 4_59 PM.webm](https://user-images.githubusercontent.com/115152460/232531010-16a40903-7223-4433-a1ba-5f9c1c7879ad.webm)

It's QA approved :heavy_check_mark: 
thank you !

zuk3975 commented:
Thanks @Progi1984 :tada:
